File: rust/src/sol.rs
//! SOL format parser.
//!
//! Anti-Idle uses the AMF0 format.
//! See https://download.macromedia.com/pub/labs/amf/amf0_spec_121207.pdf.
//! Also see https://github.com/sile/amf for an implementation.
use std::collections::HashMap;

use indexmap::IndexMap;

#[derive(Debug, Clone)]
pub enum Amf0Value {
    Number(f64),
    Boolean(bool),
    String(String),
    Null,
    Undefined,
    EcmaArray(IndexMap<usize, Amf0Value>),
}

#[derive(Debug)]
pub struct Savedata {
    pub name: String,
    pub data: HashMap<String, Amf0Value>,
}

const END_OF_OBJECT: [u8; 3] = [0x00, 0x00, 0x09];

struct Parser<'a> {
    data: &'a [u8],
    pos: usize,
}

impl<'a> Parser<'a> {
    fn new(data: &'a [u8]) -> Self {
        Self { data, pos: 0 }
    }

    fn read_bytes(&mut self, len: usize) -> &[u8] {
        if self.pos + len > self.data.len() {
            panic!("unexpected end of data");
        }
        let bytes = &self.data[self.pos..self.pos + len];
        self.pos += len;
        bytes
    }

    fn has_remaining_data(&self) -> bool {
        self.pos < self.data.len()
    }

    fn read_u8(&mut self) -> u8 {
        self.read_bytes(1)[0]
    }

    fn read_u16(&mut self) -> u16 {
        u16::from_be_bytes(self.read_bytes(2).try_into().unwrap())
    }

    fn read_u32(&mut self) -> u32 {
        u32::from_be_bytes(self.read_bytes(4).try_into().unwrap())
    }

    fn read_f64(&mut self) -> f64 {
        f64::from_be_bytes(self.read_bytes(8).try_into().unwrap())
    }

    fn read_bool(&mut self) -> bool {
        self.read_u8() != 0
    }

    fn read_str(&mut self) -> String {
        let len = self.read_u16();
        let bytes = self.read_bytes(len as usize);
        String::from_utf8(bytes.to_vec()).expect("invalid utf-8 in decoded string")
    }

    fn check_end_of_object(&mut self) -> bool {
        self.data[self.pos..self.pos + 3] == END_OF_OBJECT
    }

    fn read_end_of_object(&mut self) -> bool {
        self.read_bytes(3) == END_OF_OBJECT
    }

    fn read_key_value_pair(&mut self) -> (String, Amf0Value) {
        (self.read_str(), self.read_amf0_value())
    }

    fn read_ecma_array(&mut self) -> IndexMap<usize, Amf0Value> {
        // Skip array length
        self.read_u32();

        let mut map = IndexMap::new();
        while !self.check_end_of_object() {
            let (key, value) = self.read_key_value_pair();
            let key = key.parse::<usize>().unwrap();
            map.insert(key, value);
        }

        assert!(self.read_end_of_object());

        map
    }

    fn read_amf0_value(&mut self) -> Amf0Value {
        let tag = self.read_u8();
        match tag {
            0x0 => Amf0Value::Number(self.read_f64()),
            0x1 => Amf0Value::Boolean(self.read_bool()),
            0x2 => Amf0Value::String(self.read_str()),
            0x5 => Amf0Value::Null,
            0x6 => Amf0Value::Undefined,
            0x8 => Amf0Value::EcmaArray(self.read_ecma_array()),
            _ => panic!("unexpected tag {tag}"),
        }
    }

    fn read_header(&mut self) -> String {
        // Skip magic marker (00 BF)
        let magic = self.read_bytes(2);
        if magic != [0x00, 0xBF] {
            panic!("Invalid SOL file: wrong magic bytes");
        }

        // Length
        self.read_u32();

        // Validate TCSO magic
        if self.read_bytes(4) != b"TCSO" {
            panic!("Invalid SOL file: corrupted magic");
        }

        // Skip unknown 6 bytes
        self.read_bytes(6);

        // Read SOL name
        let sol_name = self.read_str();

        // Check version (must be 0 for AMF0)
        let version = self.read_u32();
        if version != 0 {
            panic!("Invalid SOL file: AMF0 expected, AMF3 found")
        }

        sol_name
    }
}

pub fn read_sol(data: &[u8]) -> Savedata {
    let mut parser = Parser::new(data);
    let name = parser.read_header();
    println!("Reading save from SOL, internal name: {}", name);

    let mut save = Savedata {
        name,
        data: HashMap::new(),
    };

    while parser.has_remaining_data() {
        let (key, value) = parser.read_key_value_pair();
        dbg!(&key, &value);
        save.data.insert(key, value);

        // Check trailing byte (must be 0x00)
        let trailing = parser.read_u8();
        if trailing != 0 {
            panic!("Invalid trailing byte: must be 0, got {trailing}");
        }
        dbg!(parser.pos);
    }

    save
}

#[cfg(test)]
mod test {
    use super::*;
    use std::fs::read;

    #[test]
    pub fn atg_global() {
        let file = read("tests/sol/ATG_Global.sol").unwrap();
        let save = read_sol(&file);
    }
}

File: lib/Sol.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using System.Reflection;
using System.Runtime;
using Godot;

namespace aitg.lib
{
    /// <summary>
    /// A sparse array backed by a Dictionary<int, T>,àà
    /// </summary>
    public class SparseArray<T>
    {
        private readonly Dictionary<int, T> _inner = new();

        public T this[int index]
        {
            get => _inner[index];
            set => _inner[index] = value;
        }

        public override string ToString()
        {
            var sb = new StringBuilder("{ ");
            bool first = true;
            foreach (var kvp in _inner)
            {
                if (!first) sb.Append(", ");
                sb.Append($"{kvp.Key}: {kvp.Value}");
                first = false;
            }
            sb.Append(" }");
            return sb.ToString();
        }
    }

    /// <summary>
    /// Discriminated union representing the various AMF0 value types.
    /// </summary>
    public abstract record Amf0Value
    {
        // Each possible variant is a nested record.
        // The "sealed" keyword ensures no further inheritance from each variant.

        public sealed record NumberValue(double Value) : Amf0Value;
        public sealed record BooleanValue(bool Value) : Amf0Value;
        public sealed record StringValue(string Value) : Amf0Value;
        public sealed record NullValue : Amf0Value;
        public sealed record UndefinedValue : Amf0Value;
        public sealed record EcmaArrayValue(SparseArray<Amf0Value> Value)
                                                            : Amf0Value;

        // Optionally, helper methods to construct each variant more succinctly
        public static Amf0Value Number(double value) => new NumberValue(value);
        public static Amf0Value Boolean(bool value) => new BooleanValue(value);
        public static Amf0Value String(string value) => new StringValue(value);
        public static Amf0Value Null() => new NullValue();
        public static Amf0Value Undefined() => new UndefinedValue();
        public static Amf0Value EcmaArray(SparseArray<Amf0Value> arr) => new EcmaArrayValue(arr);
    }

    /// <summary>
    /// Simple container for a SOL name and its AMF0 data.
    /// </summary>
    public sealed record SaveData(string Name, Dictionary<string, Amf0Value> Data);

    /// <summary>
    /// Internal class that parses the SOL file.
    /// </summary>
    internal class Parser
    {
        private static readonly byte[] END_OF_OBJECT = { 0x00, 0x00, 0x09 };

        private readonly byte[] _data;
        private int _pos;

        public Parser(byte[] data)
        {
            _data = data;
            _pos = 0;
        }

        private byte[] ReadBytes(int len)
        {
            if (_pos + len > _data.Length)
                throw new Exception("Unexpected end of data.");

            var result = new byte[len];
            Array.Copy(_data, _pos, result, 0, len);
            _pos += len;
            return result;
        }

        private bool HasRemainingData => _pos < _data.Length;

        private byte ReadU8() => ReadBytes(1)[0];

        private ushort ReadU16()
        {
            var bytes = ReadBytes(2);
            Array.Reverse(bytes);
            return BitConverter.ToUInt16(bytes, 0);
        }

        private uint ReadU32()
        {
            var bytes = ReadBytes(4);
            Array.Reverse(bytes);
            return BitConverter.ToUInt32(bytes, 0);
        }

        private double ReadF64()
        {
            var bytes = ReadBytes(8);
            Array.Reverse(bytes);
            return BitConverter.ToDouble(bytes, 0);
        }

        private bool ReadBool() => ReadU8() != 0;

        private string ReadStr()
        {
            var len = ReadU16();
            var bytes = ReadBytes(len);
            return Encoding.UTF8.GetString(bytes);
        }

        private bool CheckEndOfObject()
        {
            return _data[_pos] == END_OF_OBJECT[0]
                && _data[_pos + 1] == END_OF_OBJECT[1]
                && _data[_pos + 2] == END_OF_OBJECT[2];
        }

        private bool ReadEndOfObject()
        {
            var check = ReadBytes(3);
            return check[0] == END_OF_OBJECT[0]
                && check[1] == END_OF_OBJECT[1]
                && check[2] == END_OF_OBJECT[2];
        }

        private (string key, Amf0Value value) ReadKeyValuePair()
        {
            var key = ReadStr();
            var value = ReadAmf0Value();
            return (key, value);
        }

        private SparseArray<Amf0Value> ReadEcmaArray()
        {
            // Skip array length
            ReadU32();

            var map = new SparseArray<Amf0Value>();
            while (!CheckEndOfObject())
            {
                var (key, value) = ReadKeyValuePair();

                // Some real saves store "undefined" as a key - discard it
                if (key == "undefined")
                    continue;

                // Try parse the key as an index
                if (!int.TryParse(key, out int index))
                    throw new Exception($"Cannot parse '{key}' as an integer index for ECMA array.");

                map[index] = value;
            }

            if (!ReadEndOfObject())
                throw new Exception("ECMA array did not end with the expected marker.");

            return map;
        }

        private Amf0Value ReadAmf0Value()
        {
            var tag = ReadU8();
            return tag switch
            {
                0x0 => Amf0Value.Number(ReadF64()),
                0x1 => Amf0Value.Boolean(ReadBool()),
                0x2 => Amf0Value.String(ReadStr()),
                0x5 => Amf0Value.Null(),
                0x6 => Amf0Value.Undefined(),
                0x8 => Amf0Value.EcmaArray(ReadEcmaArray()),

                _ => throw new Exception($"Unexpected AMF0 tag: 0x{tag:X2}")
            };
        }

        /// <summary>
        /// Reads the SOL header data. Returns the "internal" name used in the SOL.
        /// </summary>
        public string ReadHeader()
        {
            // Skip magic marker (00 BF)
            var magic = ReadBytes(2);
            if (magic[0] != 0x00 || magic[1] != 0xBF)
                throw new Exception("Invalid SOL file: wrong magic bytes.");

            // Length (unused, but must skip)
            ReadU32();

            // Validate TCSO magic
            var tcso = ReadBytes(4);
            if (tcso[0] != 'T' || tcso[1] != 'C' || tcso[2] != 'S' || tcso[3] != 'O')
                throw new Exception("Invalid SOL file: TCSO magic not found or corrupted.");

            // Skip unknown 6 bytes
            ReadBytes(6);

            // Read SOL name
            var solName = ReadStr();

            // Check version (must be 0 for AMF0)
            var version = ReadU32();
            if (version != 0)
                throw new Exception("Invalid SOL file: AMF3 found, expected AMF0.");

            return solName;
        }

        /// <summary>
        /// Reads the file of key-value pairs until no more data remains.
        /// Each pair is followed by a trailing 0x00.
        /// </summary>
        public Dictionary<string, Amf0Value> Read()
        {
            ReadHeader();

            var dict = new Dictionary<string, Amf0Value>();

            while (HasRemainingData)
            {
                var (key, value) = ReadKeyValuePair();
                dict[key] = value;

                // Each key-value pair is followed by a trailing byte (must be 0x00)
                var trailing = ReadU8();
                if (trailing != 0)
                    throw new Exception($"Invalid trailing byte: must be 0x00, got 0x{trailing:X2}");
            }

            return dict;
        }
    }

    /// <summary>
    /// Deserializes an AMF0 value to the given class, using reflection.
    /// </summary>
    static class Amf0ReflectionDeserializer
    {

        /// <summary>
        /// Deserializes a Dictionary of (string -> Amf0Value) into a new instance of T,
        /// by matching dictionary keys to T's property names via reflection.
        /// </summary>
        public static T Deserialize<T>(Dictionary<string, Amf0Value> data)
            where T : new()
        {
            // Create a fresh instance of T, which might already have default values via property initializers
            T instance = new();

            // Iterate over all *writable* public instance properties
            var fields = typeof(T)
                .GetFields(BindingFlags.Public | BindingFlags.Instance);

            foreach (var field in fields)
            {
                var value = data[field.Name];
                object convertedValue = ConvertAmf0Value(value, field.FieldType);
                field.SetValue(instance, convertedValue);
            }

            return instance;
        }

        /// <summary>
        /// Converts an Amf0Value variant into a .NET object of the desired targetType,
        /// if possible. Returns null on unsupported conversions.
        /// </summary>
        private static object ConvertAmf0Value(Amf0Value value, Type targetType)
        {
            // A few typical conversions (expand as needed):
            switch (value)
            {
                case Amf0Value.StringValue s:
                    {
                        if (targetType == typeof(string)) return s.Value;
                        throw new Exception();
                    }

                case Amf0Value.NumberValue n:
                    {
                        // If property is double
                        if (targetType == typeof(double)) return n.Value;
                        throw new Exception();
                    }

                case Amf0Value.BooleanValue b:
                    {
                        if (targetType == typeof(bool)) return b.Value;
                        throw new Exception();
                    }

                case Amf0Value.EcmaArrayValue arr:
                    {
                        return arr;
                    }

                default:
                    // Some unknown or custom variant?
                    throw new Exception();
            }
        }
    }

    /// <summary>
    /// Public interface for reading a SOL save file.
    /// </summary>
    public static class SaveReader
    {
        /// <summary>
        /// Reads a SOL (SharedObject) file from a byte array and returns the
        /// data as a Dictionary. If the save does not exist, creates the object
        /// with default values instead.
        /// </summary>
        public static T ReadSave<T>(string saveName) where T : new()
        {
            var path = $"user://{saveName}.sol";
            if (!FileAccess.FileExists(path))
            {
                // File was not found. Return a default of the desired object
                // instead.
                return new T();
            }

            using var file = FileAccess.Open(path, FileAccess.ModeFlags.Read);
            if (file == null)
            {
                GD.PrintErr($"Failed to open file: {path}");
                throw new Exception();
            }
            var buf = file.GetBuffer((long)file.GetLength());
            var dict = new Parser(buf).Read();
            return Amf0ReflectionDeserializer.Deserialize<T>(dict);
        }
    }
}
